src/settings.cpp

```diff
@@ -840,15 +840,6 @@ directoryAddress = GetIniSetting ( IniXMLDocument, "server", "centralservaddr",
     }
     else
     {
-        // clang-format off
```

Comment:
Just removing this isn't enough to preserve behaviour.

If `servlistenabled` _exists_ and is _true_ but there's no value for `defcentservaddr`, we should set `directoryType` to `AT_DEFAULT`, I think.  (That way it's like having not changed `AT_DEFAULT` to `AT_NONE` on line 834.)

Reply:
Is there a reason to preserve such behaviour so explicitly? Would the situation you described ever occur more than once, or would the ini file written on exit no longer give that situation for subsequent runs? If so, I doubt the extra logic is worth it.

I suspect the situation described would only ever occur when the user has updated from an earlier version, the first time they run the new version. They are likely to be looking at their settings anyway, and to set them as desired.

Reply:
All of the "backwards compatibility" in settings.cpp only happens once.

Reply:
> If ( `servlistenabled` _exists_ and is _true_ ) _AND_ ( there's no value for `defcentservaddr` _OR_ `defcentservaddr` is set and true ), we should set `directoryType` to `AT_DEFAULT`, I think. (That way it's like having not changed `AT_DEFAULT` to `AT_NONE` on line 834.)

Hmm, I don't think we should do that if `defcentservaddr` is missing, only if it's set and true. `defcentservaddr` will not be in any ini file written by 3.4.7 or later.

Happy to add the logic for `serverlistenabled == true && defcentservaddr == true` to set `AT_DEFAULT`. It would only be triggered for someone upgrading from 3.4.6 or earlier directly to 3.9.0 (if such a rare person were to exist).

Reply:
Indeed rare.  Server settings are only loaded when there's a UI and there are precious few Server UI cases from back then that are listed (zero, I'd predict).  Perhaps we should announce that a two step upgrade is recommended from cases older than 3.8.2 - upgrade to 3.8.2 and then the next release, and drop compatibility with older cases.  It would clean up settings.cpp a lot.

(Of course, we can't do that for the protocol stuff so easily.)

Reply:
> Indeed rare. Server settings are only loaded when there's a UI and there are precious few Server UI cases from back then that are listed (zero, I'd predict). Perhaps we should announce that a two step upgrade is recommended from cases older than 3.8.2 - upgrade to 3.8.2 and then the next release, and drop compatibility with older cases. It would clean up settings.cpp a lot.

I think that would be good. My immediate concern is whether I need any changes to this PR to get merged. Do you have more work on settings in your NBT?

Reply:
OK, I've added the backward compatibility for < 3.4.7 by checking for `defcentservaddr` first if it exists, and setting to `AT_DEFAULT` or `AT_CUSTOM` as appropriate, and also for < 3.9.0 by testing for `servlistenabled` exists and is false _after_ doing the other settings from the ini file (but not if using the command line). I'm pretty sure this gives proper backward compatibility for all versions.

Reply:
Yep, pretty sure that should do it.